### PR TITLE
IS-719: prep replace on term

### PR DESCRIPTION
**Pull request by leeheonseung**

### CODE REVIEW

File: iconservice/icon_service_engine.py [hunk 1]
```diff
@@ -591,13 +537,15 @@ def _update_productivity(context: 'IconScoreContext',
...
-            is_validate: bool = main_prep.address in validates
-            prep: 'PRep' = context.get_prep(main_prep.address, mutable=True)
-            if prep:
-                prep.update_productivity(is_validate)
-                context.put_dirty_prep(prep)
+        main_preps: List['Address'] = context.storage.meta.get_last_main_preps(context)
+
+        for address in main_preps:
+            is_validate: bool = address in validates
+            dirty_prep: 'PRep' = context.get_prep(address, mutable=True)
```
Comment (soobokjin) [inline iconservice/icon_service_engine.py]: I guess None type data could be returned from the get_prep. (should change the type hint to Optional)

File: iconservice/icon_constant.py [hunk 2]
```diff
@@ -252,6 +252,7 @@ class DeployState(IntEnum):
 PENALTY_GRACE_PERIOD = IISS_DAY_BLOCK * 2
 
 MIN_PRODUCTIVITY_PERCENTAGE = 85
+MAX_UNVALIDATED_SEQUENCE_BLOCKS = 660
```
[the remark this answers is not in the document]
Comment (leeheonseung) [reply iconservice/icon_constant.py]: not now, but I think QA needs it for test.

File: iconservice/icon_service_engine.py [hunk 3]
```diff
@@ -753,13 +697,45 @@ def _process_base_transaction(self,
...
+            cls,
+            context: 'IconScoreContext'):
+        """Check the P-Reps to impose penalty on every block
+
+        :param context:
+        :return:
+        """
+
+        # assign
+        block_validation: List['Address'] = []
```
[the remark this answers is not in the document]
Comment (leeheonseung) [reply iconservice/icon_service_engine.py]: come to think of it, these list don't need anymore, I'll remove it.

File: iconservice/icon_service_engine.py [hunk 4]
```diff
@@ -753,13 +697,45 @@ def _process_base_transaction(self,
...
+
+        for main_prep in context.engine.prep.term.main_preps:
+            prep: 'PRep' = context.get_prep(main_prep.address)
+            assert prep is not None
+
+            if prep.is_low_productivity():
+                low_productivity.append(prep.address)
+
+            if prep.is_over_unvalidated_sequence_blocks():
+                block_validation.append(prep.address)
```
[the remark this answers is not in the document]
Comment (leeheonseung) [reply iconservice/icon_service_engine.py]: first:
    status: suspend
    eventlog: O

second:
    status: disqualified
    eventlog: O

in conclusion,
    status: disqualified
    eventlog: O

File: iconservice/iconscore/icon_system_score_base.py [hunk 5]
```diff
@@ -73,4 +72,4 @@ def get_owner(self, score_address: Optional['Address']) -> Optional['Address']:
         return IconScoreContextUtil.get_owner(self._context, score_address)
 
     def unregister_prep(self, address: 'Address'):
```
[the remark this answers is not in the document]
Comment (leeheonseung) [reply iconservice/iconscore/icon_system_score_base.py]: good advice, I'll rename it

File: iconservice/iiss/engine.py [hunk 6]
```diff
@@ -595,9 +610,13 @@ def _put_header_to_rc_db(cls, context: 'IconScoreContext'):
...
     @classmethod
-    def _put_gv(cls, context: 'IconScoreContext', next_term: 'Term'):
-        current_total_supply = context.storage.icx.get_total_supply(context)
-        current_total_prep_delegated: int = context.preps.total_delegated
+    def _put_gv(cls, context: 'IconScoreContext', term: Optional['Term']):
+        if term is None:
+            current_total_supply: int = context.storage.icx.get_total_supply(context)
+            current_total_prep_delegated: int = context.preps.total_delegated
+        else:
+            current_total_supply: int = term.total_supply
```
[the remark this answers is not in the document]
Comment (leeheonseung) [reply iconservice/iiss/engine.py]: we can send governance_variable data to RC DB before decentralize.
so we need that logic

File: iconservice/iiss/engine.py [hunk 7]
```diff
@@ -595,9 +610,13 @@ def _put_header_to_rc_db(cls, context: 'IconScoreContext'):
...
+            current_total_prep_delegated: int = term.total_delegated
```
[the remark this answers is not in the document]
Comment (leeheonseung) [reply iconservice/iiss/engine.py]: we can send governance_variable data to RCDB before decentralize.
so we need that logic

File: iconservice/database/db.py [hunk 8]
```diff
@@ -221,15 +223,28 @@ def get_from_batch(self,
 
         # get value from tx_batch
         if key in tx_batch:
-            return tx_batch[key]
+            return tx_batch[key][0]
```
[the remark this answers is not in the document]
Comment (soobokjin) [reply iconservice/database/db.py]: Yes. Each data's value in tx_batch is always tuple. So I decided not to check the type of  value

File: iconservice/database/db.py [hunk 9]
```diff
@@ -248,14 +270,24 @@ def put(self,
         if context_type == IconScoreContextType.DIRECT:
             self.key_value_db.put(key, value)
         else:
-            context.tx_batch[key] = value
+            if context.tx_batch[key]:
```
[the remark this answers is not in the document]
Comment (soobokjin) [reply iconservice/database/db.py]: I have modified this code on the other branch according to your advice.

File: iconservice/database/db.py [hunk 10]
```diff
@@ -264,7 +296,9 @@ def delete(self, context: Optional['IconScoreContext'], key: bytes):
         if context_type == IconScoreContextType.DIRECT:
             self.key_value_db.delete(key)
         else:
-            context.tx_batch[key] = None
+            if context.tx_batch[key]:
```
[the remark this answers is not in the document]
Comment (soobokjin) [reply iconservice/database/db.py]: I have modified this code on the other branch according to your advice.

File: iconservice/icon_service_engine.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (goldworm-icon) [reply iconservice/icon_service_engine.py]: If low productivity penalty is already imposed on a P-Rep, sequence block validation failure penalty should be ignored.

**Pull request merged**